### ISSUE TITLE
telegram connector convert to json

### ISSUE THREAD

**Issue by dberardo** [apache/camel-kafka-connector#1082]
is there a way to convert the output of a telegram connector (and actually of any other kind of Camel connector)  into JSON?

so far i only get this kind of string output:

`IncomingMessage{messageId=id, date=2021-03-03T15:53:03Z, from=User{id=asdasd, is_bot='false', firstName='asasdasd', lastName='null', username='asdasd'}, text='😂', chat=Chat{id='asdasd', title='null', type='private', all_members_are_administrators='false'}, photo=null, video=null, audio=null, document=null, sticker=null, location=null, entities=null, caption=null, captionEntities=null, replyMarkup=null, game=null}`

and it is not possible to use it in transformers ... how can i achieve this?

**Comment by oscerd** [comment]
With 0.8.0 or 0.7.0 and with camel-jackson in the classpath of the connector try:

```
name=CamelTelegramSourceConnector
connector.class=org.apache.camel.kafkaconnector.telegram.CamelTelegramSourceConnector
tasks.max=1

key.converter=org.apache.kafka.connect.storage.StringConverter
value.converter=org.apache.kafka.connect.converters.ByteArrayConverter

camel.source.marshal=json-jackson

topics=topic_name

camel.source.endpoint.authorizationToken=<auth_token>
camel.source.path.type=bots
```

The output will be

```
./kafkacat -b localhost:9092 -t topic_name -C
{"date":1614790989,"from":{"id":xxx,"username":"xxxx","bot":false,"first_name":"xxx","last_name":"xxx","is_bot":false},"text":"ciao","chat":{"id":"xxx","title":null,"type":"private","all_members_are_administrators":false},"photo":null,"video":null,"audio":null,"document":null,"sticker":null,"entities":null,"caption":null,"game":null,"message_id":10,"location":null,"caption_entities":null,"reply_markup":null}
```

**Comment by oscerd** [comment]
0.8.0 has camel-jackson out of the box in the dependency and same for 0.7.2.

**Comment by oscerd** [comment]
The documentation about dataformat needs to be improved. You're welcome to help.

**Comment by oscerd** [comment]
Let us know if all works for you.

**Comment by dberardo** [comment]
i will give it a try tomorrow. Please keep the issue open till then

**Comment by dberardo** [comment]
hell yeah ! it worked! 

my next natural question would be: how could have found out that by myself? is there any piece of documentation you can advice, that will help me become more independent with Camel connectors / tranform / converters ?

my landing point was of course this page: https://camel.apache.org/camel-kafka-connector/latest/connectors/camel-telegram-kafka-source-connector.html

then i have tried many sources and landed onto this page, where i thought, all output types were going to be listed: https://camel.apache.org/manual/latest/type-converter.html

then i have figured out thanks to this issue, that Camel was providing some Transformers out of the box (org.apache.camel.kafkaconnector.transforms.CamelTypeConverterTransform$Value): https://github.com/apache/camel-kafka-connector/issues/188

then i gave up trying to find documentation on such transformers, and opened this issue here, which makes me look for Marshalling options on thisp age, which however does not help me further: https://camel.apache.org/manual/latest/data-format.html

**Comment by oscerd** [comment]
The only documentation for camel-kafka-connector is https://camel.apache.org/camel-kafka-connector/latest/

You landed on pages related to plain camel. Camel is the main project, ckc is a subproject based on camel components, but it's a different runtime.

All the connectors in this project are autogenerated from the plain camel components and so it's the connectors documentation (options and whatever)

We need to add a dataformat section in the ckc documentation for sure.

Data format is a camel concept while transformers and converters in the ckc context are kafka connect concept.

For ckc the only documentation is the one I linked above

**Comment by dberardo** [comment]
thanks for wrapping that up for me @oscerd , i will stay tuned for more documentation to be released. Hopefully watching this repo could be a good start. Or maybe register to any slack channel / newsletter.

i was trying to use regular confluence transformers on the marshaled output of the converter above, but i am getting some Exceptions with data formats / Structs expected, etc.

My question is now: is it possible to use transformers that expect a JSON payload after marshaling the Byte array, or is the marshaling only happening after the chain of transformers?

**Comment by oscerd** [comment]
The marshalling is happening before the ByteArray converter, so you can use a transformer after the marshaller, but you have to deal with the format after that, which should be an input stream

**Comment by dberardo** [comment]
oh i see, i think i have to review this concept then, not very familiar to me.

so you mean that the process should look like this:

input --> marshalling --> JSON --> transformer --> JSON --> (? something ?) --> Input Stream --> ByteArray converter

Is the above process correct? If so, can i simply use a JSON converter at the end, instead of a ByteArray converter?

**Comment by oscerd** [comment]
Can you please add some more information about what your transformer should do? Because it may be possible I'm didn't understand what you want achieve.

**Comment by dberardo** [comment]
sorry for the confusion then.

actually i mean doing something really simple. I want to achieve 3 things with the transformers:

1 - add a static key to the message, to identify the Bot where it comes from (i will use many bots sourcing in one topic)
2 - add a static field in the JSON payload (something like "custom-field":123)
3 - transform the payload JSON structure to extract some of the fields in a higher hierarchy level (like chat.id becomes chat_id)

these are pretty common transformations i would say, thus i thought of using InsertField, ValueToKey, and ExtractField transformers from confluence: https://docs.confluent.io/platform/current/connect/transforms/insertfield.html

**Comment by oscerd** [comment]
I guess you'll need to write a transform with all these three step. By the way, you can call the transform after the marshalling,  yes.

**Comment by dberardo** [comment]
well my idea was to chain them like 

`transforms":["instertField1","instertField2","valueToKey","extraFields1",....]`

so in this way i would not have to program my own custom transform. But apart from this implementation details, i just wanted to make sure that JSON based transforms are actually called after the marshaling, as you are suggesting.

i will give this a quick try in the next hour and if so, finally close this issue.

**Comment by oscerd** [comment]
Yes, my only doubt is the format you'll get after the marshalling phase. Maybe you'll need to convert the stream to string. But I'm not completely sure, I should try. Let me know.

**Comment by oscerd** [comment]
The chain is one possibilty. Personally I don't like the chaining, but since the transforms are already there it makes sense

**Comment by dberardo** [comment]
i prefer to do chaining at first since i am not a Java developer, later can opt for more reusable approaches.

i have tried a simple MaskField transform and i get the same error as i had before with the InsertValue transform:

![image](https://user-images.githubusercontent.com/10696700/109967478-c6405700-7cf1-11eb-94be-3dd722065d4d.png)

The connector was working fine before, i have just added this 3 linesÖ

´
"transforms": "MaskField",
"transforms.MaskField.type": "org.apache.kafka.connect.transforms.MaskField$Value",
"transforms.MaskField.fields": "string_field"
´

as suggested in here: https://docs.confluent.io/platform/current/connect/transforms/maskfield.html

**Comment by oscerd** [comment]
So, what is moving in this connector is not a struct in the kafka connect sense of struct, it is an entity equal to what the plain camel component routes. So you cannot use those transforms, except if you transform the coming data into a struct before all of this transformations. I would write a transform myself.

**Comment by dberardo** [comment]
i see. Are you aware of any transform that could handle a conversion to such kind of "struct" required by confluence transforms?

and if not, are you aware of any other transform that i could download and use to manipulate the JSON object?

otherwise i might try to get my hands on some Java code, but would really need to know where to start from, and that would take a bit ...

**Comment by oscerd** [comment]
I'm not aware of transformations for this. I can create a transformations that show how to add a field to json, if you want. Not sure if I'll have time this week though

**Comment by dberardo** [comment]
oh that would be great, and i think also useful for other new comers. 

let me know when this is happening, as i can try to setup a dev environment to compile the java code in Docker and try to compile and deploy that class on my current running connect instance.

should we stick to this issue?

**Comment by oscerd** [comment]
Let's leave this open.

**Comment by kidpollo** [comment]
I spent hours trying to figure out the right config for `camel.source.marshal` pouring into docs and code and never found out that it was `json-jackson` I still cant find that. :P
